### PR TITLE
Move pthread TLS state native code using wasm globals

### PR DESCRIPTION
**Pull request by sbc100**
This moves pthead_self and other thread state into
native code.

### CODE REVIEW

File: system/lib/pthread/pthread_self.s [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  end_function
+
+.globl _emscripten_thread_init
+_emscripten_thread_init:
+  .functype _emscripten_thread_init (i32, i32, i32) -> ()
+  local.get 0
+  global.set thread_id
+  local.get 1
+  global.set is_main_thread
+  local.get 1
```
Comment (juj) [inline system/lib/pthread/pthread_self.s:23]: `local.get 2` here?

File: emcc.py [hunk 2]
```diff
@@ -1577,9 +1577,11 @@ def filter_out_duplicate_dynamic_libs(inputs):
       # manually export them
 
       shared.Settings.EXPORTED_FUNCTIONS += [
+        '_pthread_self',
+        '__emscripten_thread_init',
         '_emscripten_get_global_libc',
         '___pthread_tsd_run_dtors',
-        'registerPthreadPtr', '_pthread_self',
+        '_pthread_self',
```
Comment (juj) [inline emcc.py:1584]: Looks like `_pthread_self` is twice in this list now?

File: src/library_pthread.js [hunk 3]
```diff
@@ -239,7 +239,7 @@ var LibraryPThread = {
       Atomics.store(HEAPU32, (threadInfoStruct + {{{ C_STRUCTS.pthread.threadStatus }}} ) >> 2, 1); // Mark the thread as no longer running.
       _emscripten_futex_wake(threadInfoStruct + {{{ C_STRUCTS.pthread.threadStatus }}}, {{{ cDefine('INT_MAX') }}}); // wake all threads
       threadInfoStruct = selfThreadId = 0; // Not hosting a pthread anymore in this worker, reset the info structures to null.
-      registerPthreadPtr(0, 0, 0); // Unregister the thread block also inside the asm.js scope.
+      __emscripten_thread_init(0, 0, 0); // Unregister the thread block also inside the asm.js scope.
```
Comment (juj) [inline src/library_pthread.js:242]: Btw while at it, can change `asm.js` -> `wasm` in the comment here and above if you'd like.

File: system/lib/pthread/pthread_self.s [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+thread_id:
+
+.globaltype is_main_thread, i32
+is_main_thread:
+
+.globaltype is_runtime_thread, i32
+is_runtime_thread:
+
+.globl pthread_self
+pthread_self:
```
Comment (juj) [inline system/lib/pthread/pthread_self.s:11]: Wait, so any function that is defined in `.s` file gets a `_` prefixed to it? So `pthread_self` here becomes `_pthread_self`?

And any functions defined in C/C++ files do not get a `_` prefix?

That is actually pretty backwards: in native compilers all assembly code is unprefixed, and all C/C++ code is prefixed with `_`. Is there a specific reason why this was flipped for Wasm? That seems a bit arbitrary.

Comment (sbc100) [reply system/lib/pthread/pthread_self.s:11]: The `_` prefix is added to all native symbols by the emscripten JS code.   We do this purely for legacy/asm.js compatability.

The actual compiler, llvm, does not emit the `_` mangling for any symbols.     I think maybe the `_` mangling is some kind of mac or windows convention.  It certainly doesn't exist on windows.

From the linker's POV there is no way to way to the difference between a C symbol and an assembler symbols.  They are all just "native" sybmols and we mangle all of them when we import them into JS.

One day we should completely remove the mangling...

Comment (sbc100) [reply system/lib/pthread/pthread_self.s:11]: I created an issue to remove this mangling completely one day: https://github.com/emscripten-core/emscripten/issues/12844

**Review by kripken — APPROVED** [review]

**Comment by kripken** [comment]
@sbc100 

This appears to have regressed something:
```
./emcc tests/hello_world.c -pthread
```
That fails from this commit, due to `error: undefined symbol: _emscripten_thread_init (referenced by $PThread__deps: ['_emscripten_thread_init','$ERRNO_CODES','emscripten_futex_wake','$killThread','$cancelThread','$cleanupThread','$PThread'], referenced by $establishStackSpace__deps: ['$PThread'], referenced by top-level compiled C/C++ code)
`.

`-s USE_PTHREADS` instead of `-pthread` works, however.

**Comment by kripken** [comment]
Oh, nm, I had to clear the cache... I forgot we don't bump the version number for things like that any more...sorry for the noise.

**Pull request merged**